// html-validate/build/rules/input-missing-label.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rule_1 = require("../rule");
class InputMissingLabel extends rule_1.Rule {
    documentation() {
        return {
            description: "Labels are associated with the input element and is required for a17y.",
            url: rule_1.ruleDocumentationUrl(__filename),
        };
    }
    setup() {
        this.on("dom:ready", (event) => {
            const root = event.document;
            for (const elem of root.getElementsByTagName("input")) {
                if (findLabelById(root, elem.id)) {
                    continue;
                }
                if (findLabelByParent(elem)) {
                    continue;
                }
                this.report(elem, "Input element does not have a label");
            }
        });
    }
}
function findLabelById(root, id) {
    if (!id)
        return null;
    return root.find((node) => node.is("label") && node.getAttributeValue("for") === id);
}
function findLabelByParent(el) {
    let cur = el.parent;
    while (cur) {
        if (cur.is("label")) {
            return cur;
        }
        cur = cur.parent;
    }
    return null;
}
module.exports = InputMissingLabel;
